README.md
CosmoABC - Likelihood free parameter estimation for cosmology
=============================================================

`CosmoABC` is a package which enables parameter inference using an Approximate Bayesian Computation (ABC) algorithm, as described in Ishida et al., 2015 [LINK].

The code was originally designed for cosmological parameter inference from galaxy clusters number counts based on Sunyaev-Zel'dovich measurements. In this context, the cosmological simulations were performed using the NumCosmo library.

Nevertheless, the user can easily take advantadge of the ABC sampler along with his/her own simulator, as well as  test personalized summary statistics and distance functions. 


Get it now!
***********

The package can be installed using the PyPI and pip::

    $ pip install CosmoABC

Or if the tarball or repository is downloaded, in the CosmoABC directory you can install and test it::

    $ pyton setup.py install
    $ python setup.py test


Examples
********

Sample input in can be found in ~CosmoABC/examples. All example files mentioned in this section are available in that directory. 

The user input file should contain all necessary variables for simulation as well as for the ABC sampler.

A simple example of user input file, using a simulator which takes 3 parameters as input (mean, std, n) from which we want to fit only two (mean, std), would look like this ::

    path_to_obs		= data.dat   	   # path to observed data 

    param_to_fit	= mean 	std	   # parameters to fit
    param_to_sim    	= mean  std  n	   # parameters needed for simulation

    mean_lim		= -10.0  10.0	   # extreme limits for parameters
    std_lim             = 0.001   3.0


    mean_prior_par 	= -1.0  3.0	    # parameters for prior distribution
    std_prior_par	= 0.001  2.0            

    mean	= 1.0			    #fiducial parameters for simulation
    std		= 0.1
    n		= 1000

    s		= 0			    # extra parameter for distance function
    epsilon1 	= 100			    # initial distance threshold for building first particle system
    M 		= 100			    # number of particles in each particle system
    delta 	= 0.2		            # convergence criteria
    qthreshold 	= 0.75			    # quantile in distance threshold 
    file_root 	= example_2par_PS	    # root to output file names 

    simulation_func   = simulation		# simulation function provided by the user
    prior_func	      = flat_prior  flat_prior  # list of prior PDF, one for each fitted parameter
    distance_func     = distance                # distance function 


User defined simulation, distance and prior functions
-----------------------------------------------------

The most important ingredients in an ABC analysis are:

* the prior probability distributions (PDF)
* the simulator
* the distance function


CosmoABC is able to handdle user defined functions for all three elements. 
You will find example files in the corresponding directory which will help you to taylor your functions for the ABC sampler. 

Built-in options for priors PDF are:

* Gaussian
* flat
* beta


Consider the <user_function_file> containing,

.. code-block:: python 

    import numpy

    def simulation( v ):
        """
        Generates a Gaussian distributed catalog.
        """

        l1 = numpy.random.normal( loc=v['mean'], scale=v['std'], size=v['n'] )
    
        return numpy.atleast_2d( l1 ).T 


    def distance( dataset1, dataset2, s1=0 ):
        """
        Calculates distance between dataset1 and dataset2.        
        """  

        t1 = abs( numpy.mean( dataset1 ) - numpy.mean( dataset2 ) )
        t2 = abs( numpy.std( dataset1 ) - numpy.std( dataset2 ) )

        return t1 + t2




The ABC sampler can be called from the command line::

    $ run_ABC.py -i <user_input_file>  -f <user_function_file>

This will run the algorithm until the convergence criteria is reached. A pdf file containing graphical representation of the results for each particle system is 
given as output. 

If the achieved result is not satisfactory and we want to run the ABC sampler beginning from the last completed particle system *N*,  we should decrese the convergene criteria `delta` in the <user_input_file> and continue the calculations from where it was stopped.  This can also be done in the command line::

    $ continue_ABC.py -i <user_input_file> -f <user_function_file> -p N


If the sampler is running and we wish to take a look in the already calculated particle systems, we can generate the corresponding plots::

    $ plot_ABC.py -i <user_input_file> -p N


It is also possible to use it interactively.

.. code-block:: python 

    from CosmoABC.priors import flat_prior
    from CosmoABC.ABC_sampler import ABC
    from CosmoABC.plots import plot_2D
    import numpy

    def simulation( v ):
        """
        Generates a Gaussian distributed catalog.
        """

        l1 = numpy.random.normal( loc=v['mean'], scale=v['std'], size=v['n'] )
    
        return numpy.atleast_2d( l1 ).T 


    def distance( dataset1, dataset2, s1=0 ):
        """
        Calculates distance between dataset1 and dataset2.        
        """  

        t1 = abs( numpy.mean( dataset1 ) - numpy.mean( dataset2 ) )
        t2 = abs( numpy.std( dataset1 ) - numpy.std( dataset2 ) )

        return t1 + t2

     
    #define fiducial model parameters
    mean = 1.0
    std  = 0.1
    v1 = {'mean': mean, 'std': std, 'n':1000 }

    #generate 'observed' catalog
    data = simulation( v1 )

    #create dictionary of required parameter values
    params = {}
    params['param_to_fit']=['mean', 'std' ]			# parameters to fit					
    params['param_lim']=[[-10, 10], [0.001, 3.0]]		# extreme limits for parameters
    params['prior_par'] = [[-1.0, 3.0], [0.001,2.0]]		# parameters for prior distribution
    params['simulation_params'] = v1				# parameters needed for simulation

    params['mean'] = mean					# fiducial parameter value
    params['std']  = std					# fiducial parameter value
    params['s']=0						# extra parameter for distance function
    params['epsilon1'] = 50.0					# initial distance threshold
    params['M'] = 100						# number of particles in each particle system
    params['delta'] =0.2					# convergence criteria
    params['qthreshold'] = 0.75					# quantile in distance threshold

    params['file_root'] = 'example_PS'				# root to output file names
    params['prior_func'] = [ flat_prior, flat_prior]		# prior functions
    params['distance_func'] = distance  			# distance functions

    #initiate ABC sampler
    sampler_ABC = ABC( dataset1=data, params=params, simulation_func=simulation, prior_func=params['prior_func'], distance_func=params['distance_func']) 

    #build first particle system
    sys1 = sampler_ABC.BuildFirstPSystem( filename=params['file_root'] + '0.dat' )

    #update particle system until convergence
    sampler_ABC.fullABC(  params['file_root'] )

    
    #plot results
    #update parameter limits for plotting

    params['param_lim'] = [[-1.0,3.0],[0.0001,0.2]]
    plot_2D( sampler_ABC.T, 'results.pdf' , params)



NumCosmo simulations
--------------------

In order to reproduce the results of Ishida *et al.* 2015, first you need to make sure the NumCosmo library is running smoothly. 
Instructions for complete instalation and tests can be found in `<http://www.nongnu.org/numcosmo/>`.



Once the simulator is installed run the complete ABC sampler + NumCosmo cluster simulations from the command line ::

    $ run_ABC_NumCosmo.py -i <user_input_file>

This will run the complete analysis presented in Ishida *et al.*, 2015 as well as produce
plots with the corresponding results.

** WARNING**  :  This might take a while! Be patient!

Analogously to what is available for the user defined simulations, we can also continue a NumCosmo calculation from particle system *N* with::

    $ continue_ABC_NumCosmo.py -i <user_input_file> -p N

If we want to run the NumCosmo simulation with a different prior or distance function, we should define it in a separated file and run::

    $ run_ABC_NumCosmo.py -i <user_input_file> -f <user_function_file>


Documentation
*************

The complete documentation can be found in [LINK].


Requirements
************

* Python 2.7
* numpy >=1.8.2
* scipy >= 0.14.0
* statsmodels >= 0.5.0
* matplotlib >= 1.3.1     
* argparse >= 1.1
* imp
* math
* argparse


Optional
--------

* NumCosmo  `<http://www.nongnu.org/numcosmo/>`


License
********

* GNU General Public License (GPL>=3)


Contents:

.. toctree::
   :maxdepth: 2



Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`


CHANGES.md
v1.0.10 Jul 12 2021 -- Update to python3.  
v1.0.9 Jan 31 2018 -- Fixed covariance call file name.  
v1.0.8 Jan 31 2018 -- Accept covariance matrix in distance test.  
v1.0.7 Nov 15 2017 -- Fixed prior call on distance test.  
v1.0.6 Sep 26 2017 -- Upload corrected version to pip.  
v1.0.5 May 20 2015 -- Fixed plotting issues and update example files.  
v1.0.4 May 19 2015 -- Fixed random seed call in priors.  
v1.0.3 May 17 2015 -- Fixed plot titles bug.  
                   -- Removed child paralelization from distance GRBF function.  
v1.0.2 May 16 2015 -- Remove wrong call for NumCosmo in plotting functions.  
v1.0.1 May 12 2015 -- Add documentation.  
v1.0.0 May 11 2015 -- First official release.  



serial_ABC.py
from math import *
from gi.repository import GObject
import matplotlib.pyplot as plt
from gi.repository import NumCosmo as Nc
from gi.repository import NumCosmoMath as Ncm

from scipy.stats.mstats import mquantiles
from scipy.stats import norm


import numpy
import random 
import os 
import shutil

from multiprocessing import Pool
import time

starttime=time.time()
############################################################
### Fiducial Cosmological parameters: ref. arXiv:1203.5775 (table 5. wCDM CMB+BAO+H0+SNeIa+SPTcl), Tgamma0 and ns are not given.
zmin = 0.3              #minimum redshift
zmax = 1.32             #maximum redshift
H0 = 71.15               #Hubble parameter
Omegam = 0.262           #Dark matter density
Omegab = 0.044          #Baryon density
Tgamma0 = 2.725          #Radiation temperature today
ns = 0.97                #spectral index 
sigma8 = 0.807           #sigma8
w = -1.01                #Dark energy equation of state     

#choose observable
observable = 'true_mass'


#mass bin
#dm = [5*10**13, 10**14, 10**14.25, 10**14.5, 10**14.75,  10**15, 10**15.25,  10**15.5, 10**15.75 ]
#dm_choose = [10**14.3, 10**14.5, 10**14.7,  10**14.9, 10**15.1, 10**15.3,  10**15.5, 10**15.7 ]

mass_min = 10**14.3
mass_max = 10**16

#quantile list
quant_list = [ 0.02, 0.09, 0.25, 0.5, 0.75, 0.91, 0.98]

#sky area
area = 2500

#######
# path and name to mock data file
mock_data = "/home/emille/Dropbox/WGC/ABC/data/Mock_Data.dat"
#mock_data = "data_fid.dat"  

Ncpu=1 # number of cpus to be used

Ninit=500 # initial number of samples 

N=250      # particle sample size after the first iteration

												
epsilon_ini = [1e20, 1e20]		#Starting tolerance

#if seed == False, use random 
seed = False  # seed for ncount



##############################################################


# Load mock data
data_fid = numpy.loadtxt( mock_data )

if observable == 'true_mass':
    dm_choose = [ numpy.log( mass_min ) ]
else:
    dm_choose = [ min( data_fid[:,1] ) ]

mq = mquantiles( data_fid[:,1], prob=quant_list )
for it in mq:
    dm_choose.append( it )

if observable == 'SZ':
    dm_choose.append( numpy.log( mass_max ) )
else:
    dm_choose.append( max( data_fid[:,1] ) ) 
#path to results directory

path1="."

#path1="/home/emille/Dropbox/WGC/ABC/RESULTS/" + observable + '/' + str( Nparams ) + 'p/om_w/'
output_param_file_root = 'SMC_ABC_' + observable + '_'  
output_covariance_file = 'covariance_evol_' + observable + '.dat'

#create output directory
if not os.path.exists( path1 ):
    os.makedirs( path1 )

#copy input file to directory for this run
#shutil.copy2( 'sequential_ABC.py', path1 + 'input.py' )   

if observable == 'true_mass':
    from functions import *
    #shutil.copy2( 'NCountSimul_true_mass_nw.py', path1 + 'functions.py' )   
else:
    from NCountSimul_SPT import *
    #shutil.copy2( 'NCountSimul_SPT.py', path1 + 'functions.py' )  


   
CosmoParams=ChooseParamsInput()

CosmoParams.params={"H0":H0,"Ob":Omegab,"Om":Omegam,"OL":1.-Omegam-Omegab,"Tgamma":Tgamma0,"ns":ns,"sigma8":0.8,"w":-1.0}

#CosmoParams.keys=["Om","w","sigma8"]
CosmoParams.keys=["Om","sigma8"]

#CosmoParams.keys_values=numpy.array( [  0.25 , -1.01,  0.7 ] )
CosmoParams.keys_values=numpy.array( [  0.25 ,  0.7 ] )

#CosmoParams.keys_cov=numpy.diag( [1.0 , 1.0, 1.0 ] )**2.
CosmoParams.keys_cov=numpy.diag( [1.0, 1.0 ] )**2.


#CosmoParams.keys_bounds=numpy.array( [ [0.0 , -10.0, 0.3 ] , [1.0-Omegab, 1.0, 1.0] ] )
CosmoParams.keys_bounds=numpy.array( [ [0.0 , 0.6 ] , [1.0-Omegab,  1.0] ] )


#desired final variance (percentage in relation to the initial variance)
#CosmoParams.desired_variance = [ 0.05, 0.05, 2.0]
#CosmoParams.desired_median = [0.05,0.05,2.0]

CosmoParams.desired_variance = [ 0.05,  0.05]
CosmoParams.desired_median = [0.05,0.05]
CosmoParams.desired_mean = [0.05,0.05]


#CosmoParams.keys=["Om"]

#CosmoParams.keys_values=numpy.array( [  0.25  ] )
#CosmoParams.keys_cov=numpy.diag( [ 0.5  ] )**2.

#CosmoParams.keys_bounds=numpy.array( [ [0.0  ] , [1.0-Omegab] ] )

CosmoParams.prior_dist="normal"

Nparams=len( CosmoParams.keys )


   
op3 = open( path1 + output_covariance_file, 'w')
for item in CosmoParams.keys:
    op3.write( item + '    ' )

op3.write( '\n' )   

epsilon = [ ]

var_flag = 0
#for ii in xrange(time_steps):
cont = 0


if Ncpu!=1:
  p=Pool(Ncpu)
else:
  pass
  

while var_flag < Nparams:

    print 'cont  = ' + str( cont )

    if ( cont == 0 ):
    
        #new simulation object
        #ncount = NCountSimul (zmin, zmax, log ( mass_min ), log ( mass_max ), area )
  

        #Generate fiducial data
        #data_fid = numpy.array( ncount.simulation( zmax, seed, CosmoParams.params)[1] )


        #if observable == 'true_mass':
        dm = dm_choose
        #else:
        #    dm = mquantiles( data_fid[:,1], prob=quant_list[1:-1] )


        #keep number of fiducial data set
        nobjs_fid = len( data_fid )

        #Calculate summary statistics for fiducial data
        summ_fid = summary_quantile( data_fid, dm, quant_list )
        
        if (Ncpu == 1):
           
           par_surv, indx,par_cov = choose_surv_par (summ_fid, dm, quant_list, epsilon_ini, Ninit, Ncpu , CosmoParams,  zmin, zmax, area,  seed, nobjs_fid, [numpy.log(mass_min), numpy.log(mass_max)])
        
        else: 
           
           job_args=[ (summ_fid, dm, quant_list, epsilon_ini, N, Ncpu , CosmoParams,  zmin, zmax, area,  seed, nobjs_fid,[numpy.log(mass_min), numpy.log(mass_max)]) for i in xrange(Ncpu)  ]
           
           results = p.map(worker, job_args)
           
           par_surv=[]
           indx=[]
           par_cov=numpy.array(results[0][-1])
           
           for i in xrange(Ncpu):
              
              
              par_surv.append(results[i][0])
           
              indx.append(results[i][1])
           
           par_surv = numpy.vstack( par_surv )
           #indx = numpy.vstack( indx ).flatten()
           indx = numpy.hstack( indx )
        
        cont = cont + 1 
        # print "Data generated"
        # print par_surv
        # print "indexes"
        # print indx
        # print "covariance"
        # print par_cov

        #sort distances according to the sum of the summary statistics distance and population tests
        final_dist = numpy.array([ sum( [ par_surv[ i ][ l1 ]/numpy.mean(par_surv[:, l1])  for l1 in xrange( -2, 0) ]) for i in range( len( par_surv ) ) ])
      
        #par_surv=par_surv[ par_surv[:,-2].argsort()  ]
      
        param_sorted = final_dist.argsort() 
      
        #par_surv=par_surv[xrange(N)]
        par_surv = numpy.array([ par_surv[ i3 ] for i3 in param_sorted[:N] ])
                 

        epsilon.append( [ mquantiles( [ par_surv[ j1 , j2] for j1 in xrange( len( par_surv ) ) ], [0.75] )[0] for j2 in range(-2,0) ]  )
        
        CosmoParams.sdata=par_surv[:]
        
        CosmoParams.sdata_weights = numpy.ones( N )/N
        
        CosmoParams.variance = [ numpy.std( CosmoParams.sdata[:, i] ) for i in range( Nparams )]
        CosmoParams.median =  [ numpy.median( CosmoParams.sdata[:, i] ) for i in range( Nparams )]
        CosmoParams.mean =  [ numpy.mean( CosmoParams.sdata[:, i] ) for i in range( Nparams )]
        print 'covariances = ' + str( CosmoParams.variance )
        print 'mean = ' + str( CosmoParams.mean )
        print 'median = ' + str( CosmoParams.median )

        for elem in CosmoParams.variance:
            op3.write( str( elem ) + '    ' )
        op3.write( '\n' )     
   

        #write results to file
        op0 = open(os.path.join( path1 , output_param_file_root + '0.dat'), 'w')
        op0.write( '#' ) 
        for par in CosmoParams.keys:
            op0.write('#' + par + '    ' )
        op0.write( 'distancia1    distancia2 \n' )
        for elem in par_surv:
           for item in elem:
              op0.write( str( item ) + '    ' )
           op0.write('\n' )
        op0.close()

      
    else:
    
        print 'cont = ' + str( cont )

        
        if (Ncpu == 1):
           
           par_surv, indx,par_cov = choose_surv_par (summ_fid, dm, quant_list, epsilon[ -1 ], N, Ncpu , CosmoParams,  zmin, zmax, area,  seed, nobjs_fid, [numpy.log(mass_min), numpy.log(mass_max)])
        
        else: 
           
           job_args=[ (summ_fid, dm, quant_list, epsilon[ -1 ], N, Ncpu , CosmoParams,  zmin, zmax, area,  seed, nobjs_fid, [numpy.log(mass_min), numpy.log(mass_max)]) for i in xrange(Ncpu)  ]
           
           results = p.map(worker, job_args)
           
           par_surv=[]
           indx=[]
           par_cov=numpy.array(results[0][-1])
           
           for i in xrange(Ncpu):
              
              
              par_surv.append(results[i][0])
           
              indx.append(results[i][1])
           
           par_surv = numpy.vstack( par_surv )
           
           #indx = numpy.vstack( indx ).flatten()
           indx = numpy.hstack( indx )
        
        
        
        CosmoParams.sdata=par_surv[:]
        
        #check if desired variance was achieved
        new_cov = [ numpy.std( CosmoParams.sdata[:, i2] ) for i2 in range( Nparams ) ]
        new_median = [ numpy.median( CosmoParams.sdata[:, i2] ) for i2 in range( Nparams ) ]
        new_mean = [ numpy.mean( CosmoParams.sdata[:, i2] ) for i2 in range( Nparams ) ]
        CosmoParams.variance_diff = [ abs( new_cov[ i2 ] - CosmoParams.variance[ i2 ] )/CosmoParams.variance[ i2 ] for i2 in range( Nparams )]
        CosmoParams.median_diff = [ abs( new_median[ i2 ] - CosmoParams.median[ i2 ] )/CosmoParams.median[ i2 ] for i2 in range( Nparams )]
        CosmoParams.mean_diff = [ abs( new_mean[ i2 ] - CosmoParams.mean[ i2 ] )/CosmoParams.mean[ i2 ] for i2 in range( Nparams )]

        var_flag = 0
        
        
        
        for j4 in range( Nparams ):
            if CosmoParams.variance_diff[ j4 ] < CosmoParams.desired_variance[ j4 ] and CosmoParams.mean_diff[ j4 ] < CosmoParams.desired_median[ j4 ]:
                var_flag = var_flag + 1
        
        print 'cov_diff = ' + str( CosmoParams.variance_diff )
        print 'median_diff  = ' + str( CosmoParams.median_diff )
        print 'mean_diff  = ' + str( CosmoParams.median_diff )
 

        for elem in CosmoParams.variance:
            op3.write( str( elem ) + '    ' )
        op3.write( '\n' )
        
        
        if var_flag < Nparams:

            CosmoParams.variance = new_cov               

            CosmoParams.median = new_median
            # store epsilon
            epsilon.append( [ mquantiles( [ par_surv[ j1 , j2 ] for j1 in xrange( len( par_surv ) ) ], [0.75] )[0] for j2 in range(-2,0) ] )
        
        
            #update the weights of the last simulation given the mean and standard deviation from the previous set of simulations
            ###########
            ####updated number of epsilon parameters     
            denominator=numpy.array([ sum([CosmoParams.sdata_weights[i]*norm_pdf_multivariate( par_surv[n,:Nparams] , par_surv[ indx[ i ] , :Nparams ],par_cov ) for i in xrange(N)]) for n in xrange(N)]	)
      
            numerator = numpy.array([ norm_pdf_multivariate(par_surv[i,:Nparams],CosmoParams.keys_values,CosmoParams.keys_cov ) for i in xrange(N)]) 
    
            weights_new= numerator / denominator
      
            weights=weights_new/weights_new.sum()
        
            CosmoParams.sdata_weights=weights[:]
            
            

        
        op1 = open(os.path.join( path1 , output_param_file_root + str( cont ) + '.dat'), 'w')
        op1.write( '#' ) 
        for par in CosmoParams.keys:
            op1.write('#' + par + '    ' )
        op1.write( 'distancia1    distancia2\n ' )
        for elem in par_surv:
           for item in elem:
              op1.write( str( item ) + '    ' )
           op1.write( '\n' )
        op1.close()

        del op1
        
        cont = cont + 1 
            
            
      
op3.close()      

if Ncpu!=1:
  p.close()
  p.join()
  
else: 
   pass
print "Done!"
print "time=%.4f seconds" % (time.time()-starttime)
